### ISSUE TITLE
Loss functions

### ISSUE THREAD

**Issue by VitorGuizilini-TRI** [MARSLab-UMN/DeepMultiviewDepth#3]
Hi, could you please also provide the loss function used for training the least-squares triangulated depth (Eq. 2)?  
Thank you very much!

**Comment by MathLens** [comment]
Sorry, I did not get your question. Eq. 2 itself is a loss function, where ground truth depth d*_i is used to supervise the optical flow output 2_u_i.

**Comment by VitorGuizilini-TRI** [comment]
Sure, I was hoping to get the pytorch code corresponding to that loss function. But I think it's already available on `run_raft.py`, I'll take a deeper look, thanks!

**Comment by MathLens** [comment]
The code has a few steps which are mathematically equivalent to eq (2) instead of using eq (2) itself, so it might take some time to follow. Let me know if you have any questions.

**Comment by VitorGuizilini-TRI** [comment]
Sure, thanks. An unit test similar to the evaluation, that takes a pre-trained model and a sample input to produce the desired loss would be very helpful.

**Comment by VitorGuizilini-TRI** [comment]
Some initial questions:

Here, `flows` starts as a N x M list of lists containing [B,2,H,W] flow estimates, where N is the number of image pairs (e.g. 5) and M is the number of RAFT iterations (e.g. 12), right? And then it's switched to M x N for the pre-triangulation step below. 
https://github.com/MARSLab-UMN/DeepMultiviewDepth/blob/d831511c9bd8cf0223063ebf2c1ac20c2aaff21d/run_raft.py#L251

Then here a residual loss for each M iteration is produced, averaging over N number of images for each iteration. 
https://github.com/MARSLab-UMN/DeepMultiviewDepth/blob/d831511c9bd8cf0223063ebf2c1ac20c2aaff21d/run_raft.py#L181

Then, in `weighted_loss` the averaged loss in each iteration is weighted using the gamma factor. One odd thing I noticed is that here the first loss is set to zero, so it doesn't count. I think it's fine, since it' the first iteration and would have a very small weight, but is that intended or should this first loss still be counted using gamma ** 11 (assuming M=12)? 

https://github.com/MARSLab-UMN/DeepMultiviewDepth/blob/d831511c9bd8cf0223063ebf2c1ac20c2aaff21d/run_raft.py#L198

**Comment by MathLens** [comment]
> Some initial questions:
> 
> Here, `flows` starts as a N x M list of lists containing [B,2,H,W] flow estimates, where N is the number of image pairs (e.g. 5) and M is the number of RAFT iterations (e.g. 12), right? And then it's switched to M x N for the pre-triangulation step below.
> https://github.com/MARSLab-UMN/DeepMultiviewDepth/blob/d831511c9bd8cf0223063ebf2c1ac20c2aaff21d/run_raft.py#L251

Right

> 
> Then here a residual loss for each M iteration is produced, averaging over N number of images for each iteration.
> https://github.com/MARSLab-UMN/DeepMultiviewDepth/blob/d831511c9bd8cf0223063ebf2c1ac20c2aaff21d/run_raft.py#L181
> 
> Then, in `weighted_loss` the averaged loss in each iteration is weighted using the gamma factor. One odd thing I noticed is that here the first loss is set to zero, so it doesn't count. I think it's fine, since it' the first iteration and would have a very small weight, but is that intended or should this first loss still be counted using gamma ** 11 (assuming M=12)?
> 
> https://github.com/MARSLab-UMN/DeepMultiviewDepth/blob/d831511c9bd8cf0223063ebf2c1ac20c2aaff21d/run_raft.py#L198

`losses_sum[key] = 0.0` is just an initialization term for the submission, not related to the the first or any other iterations. Iterations are indexed by i in the inner loop. The outer loop can be ignored here since there is only one key.
